pkg/telemetry/tracer/tracer.go

```diff
@@ -0,0 +1,40 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"context"
+
+	otrace "go.opentelemetry.io/otel/trace"
+)
+
+type tracerKeyType struct{}
+
+var tracerKey = &tracerKeyType{}
+
+// NewContext returns a derived context with the tracer set.
+func NewContext(ctx context.Context, t otrace.Tracer) context.Context {
```

Comment:
I think we should have a `NewContextWithTracer` in order to avoid this pattern of `NewContext` of `FromContext`.